apps/builder/app/builder/features/breakpoints/breakpoints-editor.tsx

```diff
@@ -1,95 +1,145 @@
-import { useState } from "react";
-import { useDebouncedCallback } from "use-debounce";
+import { Fragment, useEffect, useRef, useState } from "react";
 import { nanoid } from "nanoid";
 import store from "immerhin";
 import type { Breakpoint } from "@webstudio-is/project-build";
 import {
   theme,
-  Button,
-  DeprecatedTextField,
   Flex,
-  DeprecatedText2,
+  PanelTitle,
+  Select,
+  IconButton,
+  InputField,
+  Text,
+  PopoverSeparator,
+  Separator,
+  Box,
 } from "@webstudio-is/design-system";
-import { PlusIcon, TrashIcon } from "@webstudio-is/icons";
+import { MinusIcon, PlusIcon } from "@webstudio-is/icons";
 import { breakpointsStore } from "~/shared/nano-states";
 import { useStore } from "@nanostores/react";
+import { groupBreakpoints, isBaseBreakpoint } from "~/shared/breakpoints";
+import { z } from "zod";
 
 type BreakpointEditorItemProps = {
   breakpoint: Breakpoint;
-  onChange: (breakpoint: Breakpoint) => void;
+  autoFocus?: boolean;
+  onChangeComplete: (breakpoint: Breakpoint) => void;
   onDelete: (breakpoint: Breakpoint) => void;
 };
 
+const BreakpointFormData = z.object({
+  label: z.string(),
+  type: z.enum(["minWidth", "maxWidth"]),
+  value: z.string().transform(Number),
+});
+
+const useHandleChangeComplete = (
+  breakpoint: Breakpoint,
+  onChangeComplete: (breakpoint: Breakpoint) => void
+) => {
+  const formRef = useRef<HTMLFormElement>(null);
+  const handleChangeComplete = () => {
+    const form = formRef.current;
+    if (form === null || form.reportValidity() === false) {
+      return;
+    }
+    const parsed = BreakpointFormData.safeParse(
+      Object.fromEntries(new FormData(form))
```

Comment:
Does `Object.fromEntries(new FormData(form))` even work? Look not obvious.

Reply:
yes, its how its designed to work together, do you have a better suggestion?